rclrs/src/node/client.rs

```diff
@@ -0,0 +1,268 @@
+        let type_support = <T as rosidl_runtime_rs::Service>::get_type_support()
+            as *const rosidl_service_type_support_t;
+        let topic_c_string = CString::new(topic).map_err(|err| RclrsError::StringContainsNul {
+            err,
+            s: topic.into(),
+        })?;
+        let rcl_node = { &mut *node.rcl_node_mtx.lock() };
+
+        // SAFETY: No preconditions for this function.
+        let client_options = unsafe { rcl_client_get_default_options() };
+
+        unsafe {
+            // SAFETY: The rcl_client is zero-initialized as expected by this function.
+            // The rcl_node is kept alive because it is co-owned by the client.
+            // The topic name and the options are copied by this function, so they can be dropped
+            // afterwards.
+            rcl_client_init(
+                &mut rcl_client,
+                rcl_node,
+                type_support,
+                topic_c_string.as_ptr(),
+                &client_options,
+            )
+            .ok()?;
+        }
+
+        let handle = Arc::new(ClientHandle {
+            rcl_client_mtx: Mutex::new(rcl_client),
+            rcl_node_mtx: node.rcl_node_mtx.clone(),
+        });
+
+        Ok(Self {
+            handle,
+            requests: Mutex::new(HashMap::new()),
+            futures: Arc::new(Mutex::new(
+                HashMap::<RequestId, oneshot::Sender<T::Response>>::new(),
+            )),
+        })
+    }
+
+    /// Send a requests with a callback as a parameter.
+    ///
+    /// The [`MessageCow`] trait is implemented by any
+    /// [`Message`] as well as any reference to a `Message`.
+    ///
+    /// The reason for allowing owned messages is that publishing owned messages can be more
+    /// efficient in the case of idiomatic messages[^note].
+    ///
+    /// [^note]: See the [`Message`] trait for an explanation of "idiomatic".
+    ///
+    /// Hence, when a message will not be needed anymore after publishing, pass it by value.
+    /// When a message will be needed again after publishing, pass it by reference, instead of cloning and passing by value.
+    pub fn async_send_request_with_callback<'a, M: MessageCow<'a, T::Request>, F>(
+        &self,
+        message: M,
+        callback: F,
+    ) -> Result<(), RclrsError>
+    where
+        F: FnOnce(&T::Response) + 'static + Send,
+    {
+        let rmw_message = T::Request::into_rmw_message(message.into_cow());
+        let mut sequence_number = -1;
+        let ret = unsafe {
+            // SAFETY: The request type is guaranteed to match the client type by the type system.
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
+            )
+        };
+        let requests = &mut *self.requests.lock();
+        requests.insert(sequence_number, Box::new(callback));
+        ret.ok()
```

Comment:
The `ok()` should be called directly on the rcl function. Otherwise if there is an error, we will have this invalid entry in our requests map.

Reply:
Done in 6f2aa36